test/token/ERC20/extensions/ERC4626.test.js

```diff
@@ -38,372 +20,378 @@ contract('ERC4626', function (accounts) {
-
-      const { tx } = await this.vault.deposit(parseToken(1), recipient, { from: holder });
-
-      await expectEvent.inTransaction(tx, this.token, 'Transfer', {
-        from: holder,
-        to: this.vault.address,
-        value: parseToken(1),
-      });
-
-      await expectEvent.inTransaction(tx, this.vault, 'Transfer', {
-        from: constants.ZERO_ADDRESS,
-        to: recipient,
-        value: parseShare(100),
-      });
-    });
-
-    it('mint', async function () {
-      expect(await this.vault.maxMint(holder)).to.be.bignumber.equal(constants.MAX_UINT256);
-      expect(await this.vault.previewMint(parseShare(1))).to.be.bignumber.equal(parseToken(1).divn(100));
-
-      const { tx } = await this.vault.mint(parseShare(1), recipient, { from: holder });
-
-      await expectEvent.inTransaction(tx, this.token, 'Transfer', {
-        from: holder,
-        to: this.vault.address,
-        value: parseToken(1).divn(100),
-      });
-
-      await expectEvent.inTransaction(tx, this.vault, 'Transfer', {
-        from: constants.ZERO_ADDRESS,
-        to: recipient,
-        value: parseShare(1),
+  for (const offset of [0, 6, 18].map(web3.utils.toBN)) {
+    const parseToken = token => web3.utils.toBN(10).pow(decimals).muln(token);
+    const parseShare = share => web3.utils.toBN(10).pow(decimals.add(offset)).muln(share);
+
+    const virtualAssets = web3.utils.toBN(1);
+    const virtualShares = web3.utils.toBN(10).pow(offset);
+
+    describe(`offset: ${offset}`, function () {
+      beforeEach(async function () {
+        this.token = await ERC20Decimals.new(name, symbol, decimals);
+        this.vault = await ERC4626OffsetMock.new(name + ' Vault', symbol + 'V', this.token.address, offset);
+
+        await this.token.$_mint(holder, constants.MAX_INT256); // 50% of maximum
+        await this.token.approve(this.vault.address, constants.MAX_UINT256, { from: holder });
+        await this.vault.approve(spender, constants.MAX_UINT256, { from: holder });
+      });
+
+      it('metadata', async function () {
+        expect(await this.vault.name()).to.be.equal(name + ' Vault');
+        expect(await this.vault.symbol()).to.be.equal(symbol + 'V');
+        expect(await this.vault.decimals()).to.be.bignumber.equal(decimals.add(offset));
+        expect(await this.vault.asset()).to.be.equal(this.token.address);
+      });
+
+      describe('empty vault: no assets & no shares', function () {
+        it('status', async function () {
+          expect(await this.vault.totalAssets()).to.be.bignumber.equal('0');
+        });
+
+        it('deposit', async function () {
+          expect(await this.vault.maxDeposit(holder)).to.be.bignumber.equal(constants.MAX_UINT256);
+          expect(await this.vault.previewDeposit(parseToken(1))).to.be.bignumber.equal(parseShare(1));
+
+          const { tx } = await this.vault.deposit(parseToken(1), recipient, { from: holder });
+
+          await expectEvent.inTransaction(tx, this.token, 'Transfer', {
```

Comment:
Maybe I missed the somehow, but the tests do not include any event emission tests for `Deposit` and `Withdraw`. I would say those are the most important ones to test in this context.

Reply:
👍🏻

Reply:
Addressed here: https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4072

(you should work with us 😄 )

Reply:
highly appreciate your comment and am happy it improves the overall code quality - I can't commit anything full-time currently and will keep monitoring the OZ contracts as a side project for now (but never say never 😄)